File: rust-droid/lexer.rs
// lexer.rs

#[allow(non_camel_case_types)]
#[derive(Debug)]
pub enum TokenType {
    KEYWORD_INSERT,
    KEYWORD_UPDATE,
    KEYWORD_DELETE,
    KEYWORD_SELECT,
    KEYWORD_INTO,
    KEYWORD_VALUES,
    KEYWORD_FROM,
    KEYWORD_WHERE,
    IDENTIFIER,
    NUMBER,
    STRING,
    SYMBOL,
    EOF,
}

#[derive(Debug)]
pub struct Token {
    pub tk_type: TokenType,
    pub token: String,
}

#[derive(Debug)]
pub struct TokenList {
    pub tokens: Vec<Token>,
    pub has_error: bool,
}

impl Token {
    fn new(tk_type: TokenType, token: String) -> Token {
        return Token {
            tk_type,
            token,
        }
    }
}

fn is_symbol(c: char) -> bool {
    match c {
        '(' | ')' | ',' | ';' | '=' | '*' => true,
        _ => false,
    }
}

fn get_keyword_type(token: &str) -> TokenType {
    match token.to_uppercase().as_str() {
        "INSERT" => TokenType::KEYWORD_INSERT,
        "UPDATE" => TokenType::KEYWORD_UPDATE,
        "DELETE" => TokenType::KEYWORD_DELETE,
        "SELECT" => TokenType::KEYWORD_SELECT,
        "INTO" => TokenType::KEYWORD_INTO,
        "VALUES" => TokenType::KEYWORD_VALUES,
        "FROM" => TokenType::KEYWORD_FROM,
        "WHERE" => TokenType::KEYWORD_WHERE,
        _ => TokenType::IDENTIFIER,
    }
}

pub fn tokenize(input: &str) -> TokenList {
    let mut list: TokenList = TokenList { tokens: Vec::new(), has_error: false };

    let mut i = 0;
    let chars: Vec<char> = input.chars().collect();
    let len = chars.len();

    while i < len {
        if chars[i] == ' ' {
            i += 1;
        } else if is_symbol(chars[i]) {
            let tk = Token::new(TokenType::SYMBOL, chars[i].to_string());
            list.tokens.push(tk);
            i += 1;
        } else if chars[i].is_alphabetic() {
            let start = i;
            while i < len && (chars[i].is_alphanumeric() || chars[i] == '_') {
                i += 1;
            }
            let length = i - start;
            let token: String = chars[start..start + length].iter().collect();
            let tk_type = get_keyword_type(&token);
            let tk = Token::new(tk_type, token);
            list.tokens.push(tk);
        } else if chars[i].is_digit(10) {
            let start = i;
            while i < len && (chars[i].is_digit(10) || chars[i] == '.') {
                i += 1;
            }
            let length = i - start;
            let token: String = chars[start..start + length].iter().collect();
            let tk = Token::new(TokenType::NUMBER, token);
            list.tokens.push(tk);
        } else if chars[i] == '"' {
            let start = i;
            i += 1;
            while i < len && chars[i] != '"' {
                i += 1;
            }
            if i == len {
                println!("[ERROR:00302] Missing closing quote");
                list.has_error = true;
                return list;
            }
            let length = i - start;
            let token: String = chars[start..start + length].iter().collect();
            let tk = Token::new(TokenType::STRING, token);
            list.tokens.push(tk);
            i += 1;
        } else {
            println!("[ERROR:00303] Unexpected character: {}", chars[i]);
            list.has_error = true;
            break;
        }
    }
    return list;
}

File: c-droid/lexer.c
#include "lexer.h"
#include <string.h>
#include <stdlib.h>

static Token create_token(char c, TokenType type) {
    Token tk;
    tk.token = malloc(2 * sizeof(char));
    tk.token[0] = c;
    tk.token[1] = '\0';
    tk.type = type;
    return tk;
}


static bool is_symbol(char c) {
    switch(c) {
        case '(': 
        case ')': 
        case ',': 
        case ';':
        case '=':
        case '*':
            return true;
    }
    return false;
}

static TokenType get_keyword_type(char *token) {
    if (strcasecmp(token, "INSERT") == 0) return TOKEN_KEYWORD_INSERT;
    if (strcasecmp(token, "UPDATE") == 0) return TOKEN_KEYWORD_UPDATE;
    if (strcasecmp(token, "DELETE") == 0) return TOKEN_KEYWORD_DELETE;
    if (strcasecmp(token, "SELECT") == 0) return TOKEN_KEYWORD_SELECT;
    if (strcasecmp(token, "INTO") == 0) return TOKEN_KEYWORD_INTO;
    if (strcasecmp(token, "SET") == 0) return TOKEN_KEYWORD_SET;
    if (strcasecmp(token, "VALUES") == 0) return TOKEN_KEYWORD_VALUES;
    if (strcasecmp(token, "FROM") == 0) return TOKEN_KEYWORD_FROM;
    if (strcasecmp(token, "WHERE") == 0) return TOKEN_KEYWORD_WHERE;
    return TOKEN_IDENTIFIER;
}

TokenList tokenize(char *input) {
    TokenList list;
    list.tokens = malloc(100 * sizeof(Token));
    list.count = 0;
    list.has_error = false;

    int i = 0;
    size_t len = strlen(input);
    while (i < len) {
        if (input[i] == ' '){
            i++;
        } else if (is_symbol(input[i])){
            Token tk = create_token(input[i], TOKEN_SYMBOL);
            list.tokens[list.count] = tk;
            list.count++;
            i++;
        } else if(isalpha(input[i])) {
            int start = i;
            while(i < len && (isalnum(input[i]) || input[i] == '_')) {
                i++;
            }
            int length = i - start;
            char *token = strndup(input + start, length);
            Token tk;
            tk.token = token;
            tk.type = get_keyword_type(token);
            list.tokens[list.count] = tk;
            list.count++;
        } else if(isdigit(input[i])) {
            int start = i;
            while(i < len && (isdigit(input[i]) || input[i] == '.')) {
                i++;
            }
            int length = i - start;
            char *token = strndup(input + start, length);
            Token tk;
            tk.token = token;
            tk.type = TOKEN_NUMBER;
            list.tokens[list.count] = tk;
            list.count++;
        } else if(input[i] == '\'') {
            int start = i + 1;
            i++;
            while(i < len && input[i] != '\'') {
                i++;
            }
            if (i == len) {
                printf("[ERROR:00302] Missing closing quote\n");
                list.has_error = true;
                return list;
            }
            int length = i - start;
            char *token = strndup(input + start, length);
            Token tk;
            tk.token = token;
            tk.type = TOKEN_STRING;
            list.tokens[list.count] = tk;
            list.count++;
            i++;
        } else {
            printf("[ERROR:00303] Unexpected character: %c\n", input[i]);
            list.has_error = true;
            break;
        }
    }

    return list;
}


void free_tokens(TokenList *tokens) {
    if (tokens && tokens->tokens != NULL){
        for (int i = 0; i < tokens->count; i++) {
            if (tokens->tokens[i].token != NULL){
                free(tokens->tokens[i].token);
            }
        }
        free(tokens->tokens);
    }
    tokens->count = 0;
}

const char* token_type_to_string(TokenType type) {
    switch (type) {
        case TOKEN_KEYWORD_INSERT: return "KEYWORD_INSERT";
        case TOKEN_KEYWORD_UPDATE: return "KEYWORD_UPDATE";
        case TOKEN_KEYWORD_DELETE: return "KEYWORD_DELETE";
        case TOKEN_KEYWORD_SELECT: return "KEYWORD_SELECT";
        case TOKEN_KEYWORD_INTO:   return "KEYWORD_INTO";
        case TOKEN_KEYWORD_SET:    return "KEYWORD_SET";
        case TOKEN_KEYWORD_VALUES: return "KEYWORD_VALUES";
        case TOKEN_KEYWORD_FROM:   return "KEYWORD_FROM";
        case TOKEN_KEYWORD_WHERE:  return "KEYWORD_WHERE";
        case TOKEN_IDENTIFIER:     return "IDENTIFIER";
        case TOKEN_NUMBER:         return "NUMBER";
        case TOKEN_STRING:         return "STRING";
        case TOKEN_SYMBOL:         return "SYMBOL";
        case TOKEN_EOF:            return "EOF";
        default:                   return "UNKNOWN";
    }
}